flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriterWrapper.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.api.writer;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.runtime.io.network.buffer.BufferConsumer;
+import org.apache.flink.runtime.io.network.partition.NoOpResultPartitionConsumableNotifier;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionConsumableNotifier;
+import org.apache.flink.runtime.taskmanager.NoOpTaskActions;
+import org.apache.flink.runtime.taskmanager.TaskActions;
+
+import java.io.IOException;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A wrapper of result partition writer for handling notification of the consumable
+ * partition which is added a {@link BufferConsumer} or finished.
+ */
+public class ResultPartitionWriterWrapper {
```

Comment:
Do we need to expose this class to all touched classes? If it would implement `ResultPartitionWriter` interface, we could just wrap it once in `Task` where it is currently created. `ResultPartitionWriter` could be used everywhere because other classes are mostly interested in `ResultPartitionWriter` but not in notification specifics.

Reply:
Actually only `RecordWriter` and `Task` need touch this wrapper class for the requirement of notification in `addBufferConsumer` and `finish`. 

- Task calls the `finish` logic, so it has to maintain `ResultPartitionWriterWrapper` for this notification.

- `ResultPartitionWriterWrapper` has to be passed into `Environment` in order to be got for creating `RecordWriter`.

`Task` temporarily also maintains the previous array of `ResultPartition` which would be changed after refactoring the current `ResultPartitionWriter` interface. Do I get your points or misunderstanding?